src/ui/drawing/gdi/GDIBitmapSurface.cpp

```diff
@@ -0,0 +1,182 @@
+#include "GDIBitmapSurface.hh"
+
+#include "ui\drawing\gdi\ImageRepository.hh"
+
+namespace ra {
+namespace ui {
+namespace drawing {
+namespace gdi {
+
+void GDIAlphaBitmapSurface::FillRectangle(int nX, int nY, int nWidth, int nHeight, Color nColor)
+{
+    assert(nWidth != 0);
+    assert(nHeight != 0);
+
+    auto nStride = ra::to_signed(GetWidth());
+    auto nFirstScanline = (GetHeight() - nY - nHeight); // bitmap memory starts with the bottom scanline
+    auto pBits = m_pBits + nStride * nFirstScanline + nX;
+    if (nStride == nWidth)
+    {
+        // doing full scanlines, just bulk fill
+        auto pEnd = pBits + nWidth * nHeight;
+        do
+        {
+            *pBits++ = nColor.ARGB;
+        } while (pBits < pEnd);
+    }
+    else
+    {
+        // partial scanlines, have to fill in strips
+        while (nHeight--)
+        {
+            auto pEnd = pBits + nWidth;
+            do
+            {
+                *pBits++ = nColor.ARGB;
+            } while (pBits < pEnd);
+
+            pBits += (nStride - nWidth);
+        }
+    }
+}
+
+static inline void BlendPixel(UINT32* nTarget, UINT32 nBlend)
+{
+    const auto alpha = nBlend >> 24;
+
+    // fully transparent - do nothing
+    if (alpha == 0)
+        return;
+
+    // fully opaque, use the blend value
+    if (alpha == 255)
+    {
+        *nTarget = nBlend;
+        return;
+    }
+
+    // blend each of the RGB values based on the blend pixel's alpha value
+    // do not modify the target pixel's alpha value.
+    UINT8* pTarget = reinterpret_cast<UINT8*>(nTarget);
+    UINT8* pBlend = reinterpret_cast<UINT8*>(&nBlend);
+
+    *pTarget++ = (UINT8)(((UINT32)*pBlend++ * alpha + (UINT32)*pTarget * (256 - alpha)) / 256);
```

Comment:
Not a big deal, this can be done in another PR. We really shouldn't be using C-Style casts though.

Reply:
fixed.